### ISSUE TITLE
How to set leverage in Binance future??

### ISSUE THREAD

**Issue by sungmin9939** [ccxt/ccxt#9923]
MUST READ THIS BEFORE SUBMITTING ISSUES (read the links, then delete this message before submitting):

- https://github.com/ccxt/ccxt/wiki/FAQ
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-submit-an-issue

Make sure your local version of CCXT is up to date. Check by comparing the output of `ccxt.version` to https://github.com/ccxt/ccxt/blob/master/package.json#L3

- OS: Window10
- Programming Language version: python 3.9.7
- CCXT version: 1.55.28

I tried to set leverage of binance future trade using the code below
```
self.lever = self.binance.set_leverage(leverage=2, symbol='DOGE/USDT')
```
but what i got is...
```
ccxt.base.errors.NotSupported: binance setLeverage() supports linear and inverse contracts only
```
So i found that when i load markets and check values corresponding to 'DOGE/USDT', and I saw that linear and inverse attribute are false. How can turn value of linear into true?

**Comment by kroitor** [comment]
@sungmin9939 you should instantiate binanceusdm instead of binance if you want to trade futures:

```Python
import ccxt
from pprint import pprint

print('CCXT Version:', ccxt.__version__)


exchange = ccxt.binanceusdm({
    'apiKey': 'YOUR_API_KEY',
    'secret': 'YOUR_SECRET',
})

markets = exchange.load_markets()

# exchange.verbose = True  # uncomment for debugging purposes if necessary

response = exchange.set_leverage(2, 'DOGE/USDT')
pprint(response)
```

Let us know if that does not answer your question.

**Comment by AxelGUILLAUMET** [comment]
> @sungmin9939 you should instantiate binanceusdm instead of binance if you want to trade futures:
> 
> ```python
> import ccxt
> from pprint import pprint
> 
> print('CCXT Version:', ccxt.__version__)
> 
> 
> exchange = ccxt.binanceusdm({
>     'apiKey': 'YOUR_API_KEY',
>     'secret': 'YOUR_SECRET',
> })
> 
> markets = exchange.load_markets()
> 
> # exchange.verbose = True  # uncomment for debugging purposes if necessary
> 
> response = exchange.set_leverage(2, 'DOGE/USDT')
> pprint(response)
> ```
> 
> Let us know if that does not answer your question.

I tried and I get this error:

LeverageBuyOrder.py
CCXT version: 1.50.14
Traceback (most recent call last):
  File "Q:\Axxel\Fichiers\DropBox Local\Dropbox\Projets Actifs\Trading\BOTS\Experimentacion\Futures - Copy\BinanceFuturesLeverageBuyOrder.py", line 19, in <module>
    response = exchange1.set_leverage(2, 'ADA/USDT')
AttributeError: 'binanceusdm' object has no attribute 'set_leverage'

**Comment by kroitor** [comment]
@AxelGUILLAUMET your version of CCXT is too old, please, upgrade to 1.56+

**Comment by AxelGUILLAUMET** [comment]
> @AxelGUILLAUMET your version of CCXT is too old, please, upgrade to 1.56+

How I cna do it? Because pip install ccxt not upgrade more...

**Comment by AxelGUILLAUMET** [comment]
I updated and the codes executes BUT I have no LEVERAGE change...

```
import ccxt
import settings
import warnings

print('CCXT version:', ccxt.__version__)

warnings.filterwarnings('ignore')

exchange1 = ccxt.binanceusdm({
    "apiKey": settings.BINANCE_API_KEY_EXCHANGE1,
    "secret": settings.BINANCE_SECRET_KEY_EXCHANGE1,
    "adjustForTimeDifference": True,
})

pair = "ADA/USDT"

exchange1.load_markets()  # load markets to get the market id from a unified symbol

response = exchange1.set_leverage(2, 'ADA/USDT')
print(response)



#return exchange.fetch_balance().get(settings.TRADE_FROM).get('free')
to_use = float(exchange1.fetch_balance().get(settings.TRADE_FROM).get('free'))
price = float(exchange1.fetchTicker(pair).get('last'))

decide_position_to_use = to_use / price
quantity_to_buy = decide_position_to_use*settings.TRADE_SLIP_RANGE

order = exchange1.createOrder(pair, 'market', 'buy', quantity_to_buy, None)
print(order)
```

**Comment by kroitor** [comment]
> And I can't select the type of leverage, CROSSED or ISOLATED.

You should do that using the `set_margin_mode` method: https://github.com/ccxt/ccxt/blob/master/python/ccxt/binance.py#L4007

**Comment by AxelGUILLAUMET** [comment]
> > And I can't select the type of leverage, CROSSED or ISOLATED.
> 
> You should do that using the `set_margin_mode` method: https://github.com/ccxt/ccxt/blob/master/python/ccxt/binance.py#L4007

OK great thanks!

Now what about the leverage that doesn't work well, It can't SET to the value I want, always the default binance value...

**Comment by kroitor** [comment]
@AxelGUILLAUMET you have to first adjust the margin mode with set_margin_mode, then adjust the max leverage with set_leverage on the isolated position, then open your position and check the actual leverage (the actual leverage may be lower than the max leverage set, that will depend on your actual margin).

**Comment by AxelGUILLAUMET** [comment]
> @AxelGUILLAUMET you have to first adjust the margin mode with set_margin_mode, then adjust the max leverage with set_leverage on the isolated position, then open your position and check the actual leverage with fetch_positions (the actual leverage may be lower than the max leverage set, that will depend on your actual margin ratio).

Here is the code I have:

```
################################
# API and ORDERS configuration #
################################

# 1st Exchange

BINANCE_API_KEY_EXCHANGE1 = 'bYooI66XOU2Eo624ptIhtcjhSb3dYDCLZFzip3I0IC621JM6BAB64DJKs2j5hBsj' #Claves API
BINANCE_SECRET_KEY_EXCHANGE1 = 'ixy28okSmf0boNU3Yj3ae5UtIb4Or8yu5dmzdwKh2sk8pjj9o5VirNddlg9989kN' #Clave SECRETA API

TRADE_SLIP_RANGE = 0.85

TRADE_FROM = 'USDT'

import ccxt

print('CCXT version:', ccxt.__version__)


exchange = ccxt.binanceusdm({
    "apiKey": BINANCE_API_KEY_EXCHANGE1,
    "secret": BINANCE_SECRET_KEY_EXCHANGE1,
    "adjustForTimeDifference": True,
})

pair = "ADA/USDT"

exchange.load_markets()

# Set Magin MODE

response = exchange.set_margin_mode(pair,'ISOLATED') # ISOLATED or CROSSED
print(response)


# Set LEVERAGE

response = exchange.set_leverage(2,pair)
print(response)

# TAKE POSITION

to_use = float(exchange.fetch_balance().get(TRADE_FROM).get('free'))
price = float(exchange.fetchTicker(pair).get('last'))

decide_position_to_use = to_use / price
quantity_to_buy = decide_position_to_use*TRADE_SLIP_RANGE

order = exchange.createOrder(pair, 'market', 'buy', quantity_to_buy, None)
print(order)

```

Q:\Axxel\Fichiers\DropBox Local\Dropbox\Projets Actifs\Trading\BOTS\Experimentacion\Futures - Copy>python BinanceFuturesLeverageBuyOrder.py
CCXT version: 1.56.23
Traceback (most recent call last):
  File "C:\Program Files\Python39\lib\site-packages\ccxt\base\exchange.py", line 613, in fetch
    response.raise_for_status()
  File "C:\Program Files\Python39\lib\site-packages\requests\models.py", line 943, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://fapi.binance.com/fapi/v1/marginType

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "Q:\Axxel\Fichiers\DropBox Local\Dropbox\Projets Actifs\Trading\BOTS\Experimentacion\Futures - Copy\BinanceFuturesLeverageBuyOrder.py", line 31, in <module>
    response = exchange.set_margin_mode(pair,'ISOLATED') # ISOLATED or CROSSED
  File "C:\Program Files\Python39\lib\site-packages\ccxt\binance.py", line 4031, in set_margin_mode
    return getattr(self, method)(self.extend(request, params))
  File "C:\Program Files\Python39\lib\site-packages\ccxt\base\exchange.py", line 457, in inner
    return entry(_self, **inner_kwargs)
  File "C:\Program Files\Python39\lib\site-packages\ccxt\binance.py", line 4159, in request
    response = self.fetch2(path, api, method, params, headers, body, config, context)
  File "C:\Program Files\Python39\lib\site-packages\ccxt\base\exchange.py", line 502, in fetch2
    return self.fetch(request['url'], request['method'], request['headers'], request['body'])
  File "C:\Program Files\Python39\lib\site-packages\ccxt\base\exchange.py", line 629, in fetch
    self.handle_errors(http_status_code, http_status_text, url, method, headers, http_response, json_response, request_headers, request_body)
  File "C:\Program Files\Python39\lib\site-packages\ccxt\binance.py", line 4140, in handle_errors
    raise ExchangeError(feedback)
ccxt.base.errors.ExchangeError: binanceusdm {"code":-4046,"msg":"No need to change margin type."}

**Comment by kroitor** [comment]
@AxelGUILLAUMET 

 > ccxt.base.errors.ExchangeError: binanceusdm {"code":-4046,"msg":"No need to change margin type."}

That error means that the margin mode for ADA/USDT is already set to ISOLATED.

**Comment by AxelGUILLAUMET** [comment]
> @AxelGUILLAUMET
> 
> > ccxt.base.errors.ExchangeError: binanceusdm {"code":-4046,"msg":"No need to change margin type."}
> 
> That error means that the margin mode for ADA/USDT is already set to ISOLATED.

OK, but whhen the code follows and place the order, unfortunately the leverage I choosed before is not respected, the order appears in binance with another leverage value. When I change the leverage manually, appears this popup to confirm the change of leverage:

![image](https://user-images.githubusercontent.com/47301337/132982621-3a885c19-bd5f-4f50-81fb-e351156aa350.png)

Is with the API this popup should appear?

Now look at the previous code, the leverage I should have is x2.

![image](https://user-images.githubusercontent.com/47301337/132982755-cbfa842c-02f3-4fe2-995e-bc8de62c12dc.png)

And I have x5.

Aparently the leverage type selection worked because I changed it to CROSSED and it applied.
